source/fe/mapping_cartesian.cc

```diff
@@ -240,53 +190,135 @@ MappingCartesian<dim, spacedim>::compute_fill(
-                face_no,
-                cell->face_orientation(face_no),
-                cell->face_flip(face_no),
-                cell->face_rotation(face_no),
-                quadrature_points.size()) :
-              // called from FESubfaceValues
-              QProjector<dim>::DataSetDescriptor::subface(
-                face_no,
-                sub_no,
-                cell->face_orientation(face_no),
-                cell->face_flip(face_no),
-                cell->face_rotation(face_no),
-                quadrature_points.size(),
-                cell->subface_case(face_no))));
-
-      for (unsigned int i = 0; i < quadrature_points.size(); ++i)
-        {
-          quadrature_points[i] = start;
-          for (unsigned int d = 0; d < dim; ++d)
-            quadrature_points[i](d) +=
-              data.cell_extents[d] * data.quadrature_points[i + offset](d);
-        }
+        QProjector<dim>::DataSetDescriptor::cell();
+
+      transform_quadrature_points(cell, data, offset, quadrature_points);
+    }
+}
+
+
+
+template <int dim, int spacedim>
+void
+MappingCartesian<dim, spacedim>::maybe_update_face_quadrature_points(
+  const typename Triangulation<dim, spacedim>::cell_iterator &cell,
+  const unsigned int                                          face_no,
+  const InternalData &                                        data,
+  std::vector<Point<dim>> &quadrature_points) const
+{
+  AssertIndexRange(face_no, GeometryInfo<dim>::faces_per_cell);
+
+  if (data.update_each & update_quadrature_points)
+    {
+      const typename QProjector<dim>::DataSetDescriptor offset =
+        QProjector<dim>::DataSetDescriptor::face(face_no,
+                                                 cell->face_orientation(
+                                                   face_no),
+                                                 cell->face_flip(face_no),
+                                                 cell->face_rotation(face_no),
+                                                 quadrature_points.size());
+
+
+      transform_quadrature_points(cell, data, offset, quadrature_points);
     }
+}
+
 
 
+template <int dim, int spacedim>
+void
+MappingCartesian<dim, spacedim>::maybe_update_subface_quadrature_points(
+  const typename Triangulation<dim, spacedim>::cell_iterator &cell,
+  const unsigned int                                          face_no,
+  const unsigned int                                          sub_no,
+  const InternalData &                                        data,
+  std::vector<Point<dim>> &quadrature_points) const
+{
+  AssertIndexRange(face_no, GeometryInfo<dim>::faces_per_cell);
+  AssertIndexRange(sub_no, GeometryInfo<dim>::max_children_per_face);
+
+  // We would like to check for
+  // sub_no < cell->face(face_no)->n_children(),
+  // but unfortunately the current
+  // function is also called for
+  // faces without children (see
+  // tests/fe/mapping.cc). Therefore,
+  // we must use following workaround.
```

Comment:
I'm not sure that I understand you what you mean, but if I change the asserts so that this is checked everytime:
```c++
  AssertIndexRange(sub_no, cell->face(face_no)->n_children());
```
then the following tests fails:

mappings/fe_face_values_1d_mapping_cartesian.debug
mappings/mapping.debug

because `reinit` is called with all subface indices from `[0, max_children_per_face)` in both of them:

line 157 in tests/mappings/mapping.cc
line 118 in tests/mappings/fe_face_values_1d_mapping_cartesian.cc

Both tests use an unrefined hypercube triangulation.

Reply:
You're right. I think (I should check) that the test verifies that we can compute subface quadratures on faces without subfaces. We should document this is a feature instead of writing this comment. I'll take a closer look later (its not immediately relevant to this refactoring PR).